### DISCUSSION TITLE
Swap x-mean and y-mean for the linear model example, Recipe #3

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [EvaMaeRey/easy-geom-recipes#11] [Announcements]
Demo  

compute_group_xmean <- function(data, scales){

data |> 
        summarize(mean_x = mean(x),
                            xintercept = mean_x)

} 

StatXMean <- ggproto("StatXMean", Stat, 
                                        compute_group = compute_group_xmean,
                                        default_aes = aes(y = I(.5),
                                                                        x = after_stat(mean_x),
                                                                        label = round(after_stat(mean_x), 2)), 
                                        dropped_aes = c("y", "x"))

Upsides: Conceptually simple, and I see people using mean line at y very often (linear model stuff - orignial Recipe #3 - may resonate with smaller audience).  teaches default_aes and after_stat, makes stat nicely usable with GeomVline and GeomLabel

Downsides:  and I() might be one thing too many that's unfamiliar?... Can't use geom_vline(stat = StatXmean) to test/use stat (stat isn't an argument in geom_vlines - which it were!), maybe dropped aes is required so you don't get warnings, and that's like one more thing that unfamiliar... though I'm not 100%...

**Comment by EvaMaeRey** [comment]
What actually works, doesn't use dropped_aes, and uses after_stat less: 

compute_group_xmean <- function(data, scales){

data |>
summarize(mean_x = mean(x),
          xintercept = mean_x, 
          y = I(.5),
          x = mean_x)

}

StatXMean <- ggproto("StatXMean", Stat,
                     compute_group = compute_group_xmean,
                     default_aes = aes(label = round(after_stat(mean_x), 2)),
                       required_aes = "x")

cars|>
  ggplot() + 
  aes(speed, dist) +
  geom_point() + 
  layer(geom = GeomVline, stat = StatXMean, position = "identity") +
  geom_label(stat = StatXMean)


Kind of surprised that following doesn't: 


``` r
library(tidyverse)
compute_group_xmean <- function(data, scales){

data |>
summarize(mean_x = mean(x),
          xintercept = mean_x, 
          mean_range_y = mean(range(y)))

}

StatXMean <- ggproto("StatXMean", Stat,
                     compute_group = compute_group_xmean,
                     default_aes = aes(label = round(after_stat(mean_x), 2),
                                       y = I(.5),
                                       x = after_stat(mean_x)))

cars|>
  ggplot() + 
  aes(speed, dist) +
  geom_point() + 
  layer(geom = GeomVline, stat = StatXMean, position = "identity") +
  geom_label(stat = StatXMean)
#> Warning: The following aesthetics were dropped during statistical transformation: x and
#> y.
#> ℹ This can happen when ggplot fails to infer the correct grouping structure in
#>   the data.
#> ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
#>   variable into a factor?
#> The following aesthetics were dropped during statistical transformation: x and
#> y.
#> ℹ This can happen when ggplot fails to infer the correct grouping structure in
#>   the data.
#> ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
#>   variable into a factor?
#> Error in `geom_label()`:
#> ! Problem while setting up geom.
#> ℹ Error occurred in the 3rd layer.
#> Caused by error in `compute_geom_1()`:
#> ! `geom_label()` requires the following missing aesthetics: x and y.
```

<sup>Created on 2024-09-30 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
There's a hint in the error message: the geom in the third layer (`GeomLabel`) is expecting the layer data handed off from the stat to include "x", "y", and "label", of which "x" and "y" are missing from the layer data at that stage:

```r
GeomLabel$required_aes
#> [1] "x"     "y"     "label"
```

So if you look at the "before geom" snapshot of the data where that hand-off happens, it is indeed the case that in the third layer `geom_label(stat = StatXMean)`, `StatXMean` has dropped the `x` and `y` columns that the `GeomLabel` needs:

```r
suppressWarnings(
  ggtrace::layer_before_geom(i = 3, error = TRUE, verbose = FALSE)
)
#> # A tibble: 1 × 6
#>   label mean_x xintercept mean_range_y PANEL group
#>   <dbl>  <dbl>      <dbl>        <dbl> <fct> <int>
#> 1  15.4   15.4       15.4           61 1        -1
```

**Reply by yjunechoe** [reply]
Elaborating, I think it's probably never a good idea to flat-out drop aesthetics/columns in Stat compute that the Geom will need later too.

Going off of the current problematic code, if we wanted to suppress all errors and warnings in the `geom_label()` layer as is, the user would need to make a lot of effort to make all the pieces align within the layer via `stage()`. Additionally `StatXMean` would _still_ have to mark `x` and `y` as aesthetics that it "drops" to avoid the warnings, since the check for that happens _right after_ compute (even if you add those columns back via `after_stat()`).

```r
library(ggplot2)
library(dplyr)

compute_group_xmean <- function(data, scales){
  
  data |>
    summarize(mean_x = mean(x),
              xintercept = mean_x, 
              mean_range_y = mean(range(y)))
  
}

StatXMean <- ggproto("StatXMean", Stat,
                     compute_group = compute_group_xmean,
                     default_aes = aes(label = round(after_stat(mean_x), 2),
                                       y = I(.5),
                                       x = after_stat(mean_x)),
                     dropped_aes = c("x", "y"))

cars|>
  ggplot() + 
  aes(speed, dist) +
  geom_point() + 
  layer(geom = GeomVline, stat = StatXMean, position = "identity") +
  geom_label(
    stat = StatXMean,
    aes(
      x = stage(speed, after_stat = mean_x),
      y = stage(start = dist, after_stat = mean_range_y)
    )
  )
```

![image](https://github.com/user-attachments/assets/3f754594-f9e5-44b3-97e2-d243fd1a5926)

I think we all agree that the above code looks pretty terrible - users should never have to do something like this just to get StatXMean and GeomLabel work together within a layer!

Not that you'd advocate for this, but in any case, having *default* delayed aesthetics can get very tricky due to these fragile things that you can easily break in the internals 😅. I think a good first step to a principled fix is to ensure that you preserve the `x` column in the stat, and map `y` to `mean_x` in the after_stat of the geom label layer to satisfy the geom.

**Reply by yjunechoe** [reply]
Also, something for the future: you might want to account for cases where the user decides to flip the orientation of a layer. So for example in StatCount$compute group, the default aes is:

```r
StatCount$default_aes
#> Aesthetic mapping: 
#> * `x`      -> `after_stat(count)`
#> * `y`      -> `after_stat(count)`
#> * `weight` -> 1
```

Which lets it flexibly "count" by x or by y simply on the basis of the aesthetic that's left unmapped by the user.

In a similar vein, it may be worth considering something like:

```r
StatMean$default_aes
#> Aesthetic mapping: 
#> * `x`      -> `after_stat(mean)`
#> * `y`      -> `after_stat(mean)`
```

Versus two separate `StatXMean` and `StatYMean` - I think(?) a single `StatMean` should be doable
